qiskit/circuit/parameterexpression.py

```diff
@@ -261,7 +264,12 @@ def _apply_operation(
         else:
             expr = operation(self_expr, other_expr)
 
-        return ParameterExpression(parameter_symbols, expr)
+        out_expr = ParameterExpression(parameter_symbols, expr)
+        out_expr._names = self._names
+        if isinstance(other, ParameterExpression):
+            out_expr._names.update(other._names)
```

Comment:
Is there a risk of sharing `._names` across instances here, without a copy? I took a quick look, and I could only find cases where `._names` was created once and not updated, but it looks like L270 above would inadvertently also update `self._names`.

Reply:
Good call, there is no risk of that in the examples I was testing (and potentially in general because anything that changes the parameters set typically involves creating a new parameter expression), but if something were to modify the names dict of self or other it would be reflected in out's name dict. It's better to be safe here and just use `dict.copy()`. Done in: https://github.com/Qiskit/qiskit-terra/pull/7031/commits/a2b5cb935142d7f440520b594a19b13c55e3d80a